pkg/registry/core/service/ipallocator/controller/repair.go

```diff
@@ -313,3 +331,69 @@ func (c *Repair) checkLeaked(leaks map[string]int, stored ipallocator.Interface,
 		}
 	})
 }
+
+// used to generate empty patch bytes
+type emptyType struct {
+}
+
+// Services with "PreferDualStack" would be
+// -- set to dual-stack on a dual-stack cluster.
+// -- set to single stack on a single stack cluster.
+// Because the way allocator work (on Create/Update) they won't be upgraded
+// downgraded (from/to dual-stack) as cluster config change. More over
+// they will be arbitrary upgraded/downgraded if the user tried to
+// update them (e.g. setting a label).
+// This func ensures that all services of policy "PreferDualStack" is aligned
+// to cluster config without the user needing to update them
+func (c *Repair) fixPreferDualStackServices() error {
+	isClusterDualStack := len(c.allocatorByFamily) == 2
+
+	svcList, err := c.serviceClient.Services(metav1.NamespaceAll).List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return fmt.Errorf("preferDualStackFixer: unable to get list of services with err: %v", err)
+	}
+
+	for _, svc := range svcList.Items {
+		if svc.Spec.Type != v1.ServiceTypeClusterIP ||
```

Comment:
NodePorts and LoadBalancer Types have ClusterIPs too, I think those should be upgraded too